### PR TITLE
[REF-6034] MLOps: Adding **kwargs To MLOps.set_stat

### PR DESCRIPTION
**Pull request by harshilpm**
- We are going to support metrics for models. In this commit, api is not accepting extra **kwargs to get extra argument
- Extra argument helps in outputting stat correctly. For example, confusion matrixx needs labels as extra argument.
- We are starting with confusion matrix (classification)

### CODE REVIEW

File: mlops/setup.py [hunk 1]
```diff
@@ -42,7 +42,7 @@
         "kazoo",
         "protobuf",
         "requests",
-        "py4j"
+        "py4j", 'scikit-learn'
```
Comment (lior-amar-pm) [inline mlops/setup.py]: Keep each requirement in a separate line

File: mlops/parallelm/mlops/stats/stats_helper.py [hunk 2]
```diff
@@ -29,23 +32,62 @@ def _validate_supported_conf_ts_data_type(self, data):
...
+            self.set_stat(name=name,
+                          data=mlops_stat_object,
+                          model_id=model_id,
+                          # type of stat will be General
+                          category=StatCategory.GENERAL,
+                          timestamp=timestamp, **kwargs)
+        else:
+            self._logger.error(
+                "{} predefined stat cannot be output as error happened in creating mlops stat object from {}"
+                    .format(name, data))
```
Comment (lior-amar-pm) [inline mlops/parallelm/mlops/stats/stats_helper.py]: Why an exception is not thrown?

File: mlops/tests/test_mlops.py [hunk 3]
```diff
@@ -294,6 +296,31 @@ def test_general_graph():
...
+
+    # second way
+    pm.metrics.confusion_matrix(y_true=labels_actual, y_pred=labels_pred, labels=labels_ordered)
+
+    # should throw error if labels are not provided
+    with pytest.raises(MLOpsException):
+        pm.set_stat(ClassificationMetrics.CONFUSION_MATRIX, cm)
+
+    with pytest.raises(MLOpsException):
+        pm.metrics.confusion_matrix(y_true=labels_actual, y_pred=labels_pred, labels=None)
```
Comment (lior-amar-pm) [inline mlops/tests/test_mlops.py:319]: Also add checks when the labels_pred and labels_actual are not of the same size.

**Review by lior-amar-pm — APPROVED** [review]

File: mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+from parallelm.mlops.stats.table import Table
+
+
+class ConfusionMatrix(object):
+    """
+    Responsibility of this class is basically creating table object out of input array and list of labels.
+    """
+
+    @staticmethod
+    def _get_mlops_cm_table_object(cm_nd_array, **kwargs):
```
Comment (sriram109) [inline mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py]: do we need to keep this a private method?

File: mlops/parallelm/mlops/mlops.py [hunk 5]
```diff
@@ -385,18 +388,51 @@ def current_model(self):
...
+        >>> cm_matrix = Table().name("User Given Confusion Matrix").cols(labels_string)
+        >>> for index in range(len(cm)):
+        >>>     cm_matrix.add_row(labels_string[index], list(cm[index]))
+        >>> mlops.set_stat(cm_matrix)
+
+        >>> ## OR - Second Way ##
+        >>> mlops.set_stat(ClassificationMetrics.CONFUSION_MATRIX, cm, labels=labels_ordered)
+
+        >>> ## OR - Third Way ##
+        >>> mlops.metrics.confusion_matrix(y_true=labels_actual, y_pred=labels_pred, labels=labels_ordered)
```
Comment (sriram109) [inline mlops/parallelm/mlops/mlops.py:428]: why is the 3rd way necessary?

File: mlops/setup.py [hunk 6]
```diff
@@ -42,7 +44,8 @@
         "kazoo",
         "protobuf",
         "requests",
-        "py4j"
+        "py4j",
+        'scikit-learn'
```
Comment (sriram109) [inline mlops/setup.py]: double quotes? for consistency

File: mlops/parallelm/mlops/mlops.py (hunk 5 above)
Comment (harshilpm) [reply mlops/parallelm/mlops/mlops.py:428]: @lior-amar-pm  wants third api as well

File: mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py (hunk 4 above)
Comment (harshilpm) [reply mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py]: i do not want it to be used without knowing what it does. User outside mlops should not use this at all. Thus.

File: mlops/setup.py (hunk 6 above)
Comment (harshilpm) [reply mlops/setup.py]: True!

File: mlops/parallelm/mlops/mlops.py (hunk 5 above)
Comment (harshilpm) [reply mlops/parallelm/mlops/mlops.py:428]: @lior-amar-pm can you please comment?

Comment (sriram109) [inline mlops/parallelm/mlops/mlops.py:428]: @lior-amar-pm can you describe why we need this 3rd way of accessing stats? i dont fully understand its need - it feels clunky

File: mlops/parallelm/mlops/channels/mlops_python_channel.py [hunk 7]
```diff
@@ -28,15 +28,16 @@ def done(self):
         pass
 
     def stat(self, name, data, model_id, category=None, model=None, model_stat=None):
-
         if category in (StatCategory.CONFIG, StatCategory.TIME_SERIES):
+
```
Comment (yakov-g) [inline mlops/parallelm/mlops/channels/mlops_python_channel.py]: please remove all newly added empty lines

File: mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+                    labels_string = [str(i) for i in labels]
+                    cm_matrix = Table().name(ClassificationMetrics.CONFUSION_MATRIX.value).cols(labels_string)
+
+                    for index in range(len(cm_nd_array)):
+                        cm_matrix.add_row(labels_string[index], list(cm_nd_array[index]))
+
+                    return cm_matrix
+
+                else:
+                    raise MLOpsException(
```
Comment (yakov-g) [inline mlops/parallelm/mlops/ml_metrics_stat/confusion_matrix.py]: I think it is time to define more specific exception name for Statistics, like MLOpsStatisticsException

or even MLOpsStatisticsClassificationMetricsException

File: mlops/parallelm/mlops/mlops.py (hunk 5 above)
Comment (lior-amar-pm) [reply mlops/parallelm/mlops/mlops.py:428]: The third way is syntactic sugar on top of the first way. 
It makes it easier to the user to call this method and to get documentation about the specific params in their development environment (e.g. pycharm, intellij..)

Comment (harshilpm) [reply mlops/parallelm/mlops/mlops.py:428]: @sriram109

**Review by sriram109 — APPROVED** [review]

**Pull request merged**